test/libsolidity/semanticTests/structs/struct_copy.sol

```diff
@@ -38,12 +38,12 @@ contract c {
 // compileViaYul: also
 // ----
 // set(uint256): 7 -> true
-// gas irOptimized: 110011
+// gas irOptimized: 110119
```

Comment:
Seems to be working 😅

Reply:
:shrug:

Reply:
The assembly is actually shorter. My guess is that the removed mstore instructions makes it impossible for the opcode-based optimizer to remove the multiple keccak calls...

Reply:
```diff
--- /tmp/ir_old.yul	2022-01-03 15:46:30.242337855 +0100
+++ /tmp/ir_opt.yul	2022-01-03 15:41:35.197221133 +0100
@@ -79,34 +79,26 @@
                 mstore(_1, var_k)
                 mstore(0x20, _1)
                 sstore(keccak256(_1, 0x40), 0x01)
-                mstore(_1, var_k)
-                mstore(0x20, _1)
                 sstore(add(keccak256(_1, 0x40), 0x01), 0x03)
-                mstore(_1, var_k)
-                mstore(0x20, _1)
                 sstore(add(keccak256(_1, 0x40), 2), 0x04)
-                mstore(_1, var_k)
-                mstore(0x20, _1)
                 sstore(add(keccak256(_1, 0x40), 0x03), 2)
                 var := 0x01
             }
             function fun_copy(var_from, var_to) -> var_
             {
-                let _1 := 0x00
-                mstore(_1, var_from)
-                mstore(0x20, _1)
-                let dataSlot := keccak256(_1, 0x40)
-                mstore(_1, var_to)
-                mstore(0x20, _1)
-                let dataSlot_1 := keccak256(_1, 0x40)
+                mstore(0x00, var_from)
+                mstore(0x20, 0x00)
+                let dataSlot := keccak256(0x00, 0x40)
+                mstore(0x00, var_to)
+                let dataSlot_1 := keccak256(0x00, 0x40)
                 if iszero(eq(dataSlot_1, dataSlot))
                 {
                     sstore(dataSlot_1, sload(dataSlot))
                     let memberSlot := add(dataSlot_1, 1)
-                    let _2 := add(dataSlot, 1)
-                    if iszero(eq(memberSlot, _2))
+                    let _1 := add(dataSlot, 1)
+                    if iszero(eq(memberSlot, _1))
                     {
-                        sstore(memberSlot, sload(_2))
+                        sstore(memberSlot, sload(_1))
                         sstore(add(dataSlot_1, 2), sload(add(dataSlot, 2)))
                     }
                     sstore(add(dataSlot_1, 3), sload(add(dataSlot, 3)))
@@ -119,18 +111,12 @@
                 mstore(_1, var_k)
                 mstore(0x20, _1)
                 var_a := sload(keccak256(_1, 0x40))
-                mstore(_1, var_k)
-                mstore(0x20, _1)
                 var_x := sload(add(keccak256(_1, 0x40), 1))
-                mstore(_1, var_k)
-                mstore(0x20, _1)
                 var_y := sload(add(keccak256(_1, 0x40), 2))
-                mstore(_1, var_k)
-                mstore(0x20, _1)
                 var_c := sload(add(keccak256(_1, 0x40), 3))
             }
         }
-        data ".metadata" hex"a36469706673582212201575720a9612338487e4ea44beef0f99559f8ddb767d3b3a8f88b30e8203f9ee6c6578706572696d656e74616cf564736f6c634300080c0041"
+        data ".metadata" hex"a36469706673582212201300a8828ade0dfe77d7dbd65f00b07037165a285fa1ba1a0c986a59b74ec9b66c6578706572696d656e74616cf564736f6c634300080c0041"
     }
 }
```

Reply:
The main change is that zero constants are now inlined, so I would say it's not an issue.

Reply:
Should perhaps implement evaluating `keccak` for multi words (at least 2 words).